skbio/stats/distance/_cutils.pyx

```diff
@@ -293,3 +293,119 @@ def permanova_f_stat_sW_cy(TReal[:, ::1] distance_matrix,
+
+cimport numpy as cnp
+
+ctypedef fused floating:
+    cnp.float32_t
+    cnp.float64_t
+
+ctypedef cnp.float32_t float32_t
+ctypedef cnp.float64_t float64_t
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def geomedian_axis_one(floating[:, :] X, floating eps=1e-7,
+                           size_t maxiters=500):
+    """Compute high dimensional median."""
+    cdef size_t p = X.shape[0]
+    cdef size_t n = X.shape[1]
+
+    cdef floating[:] y = np.mean(X, axis=1)
+
+    if n == 1:
+        return y
+
+    if floating is cnp.float32_t:
+        dtype = np.float32
+    else:
+        dtype = np.float64
+
+    cdef floating[:] D = np.empty(n, dtype=dtype)
+    cdef floating[:] Dinv = np.empty(n, dtype=dtype)
+    cdef floating[:] W = np.empty(n, dtype=dtype)
+    cdef floating[:] T = np.empty(p, dtype=dtype)
+    cdef floating[:] y1 = np.empty(p, dtype=dtype)
+    cdef floating[:] R = np.empty(p, dtype=dtype)
+
+    cdef floating dist, Dinvs, total, r, rinv, tmp, Di
+    cdef size_t nzeros = n
+    cdef size_t iteration
+
+    with nogil:
+        iteration = 0
+        while iteration < maxiters:
+
+            for i in range(n):
+                Di = dist_euclidean(X[:, i], y)
+                D[i] = Di
+                if fabs(Di) > eps:
+                    Dinv[i] = 1. / Di
+                else:
+                    Dinv[i] = 0.
+
+            Dinvs = sum(Dinv)
+
+            for i in range(n):
+                W[i] = Dinv[i] / Dinvs
+
+            for j in range(p):
+                total = 0.
+                for i in range(n):
+                    if fabs(D[i]) > eps:
+                        total += W[i] * X[j, i]
+                T[j] = total
+
+            nzeros = n
+            for i in range(n):
+                if fabs(D[i]) > eps:
+                    nzeros -= 1
+
+            if nzeros == 0:
+                y1 = T
+            elif nzeros == n:
+                break
+            else:
+                for j in range(p):
+                    R[j] = (T[j] - y[j]) * Dinvs
+                r = norm_euclidean(R)
+                if r > eps:
+                    rinv = nzeros/r
+                else:
+                    rinv = 0.
+                for j in range(p):
+                    y1[j] = max(0, 1-rinv)*T[j] + min(1, rinv)*y[j]
+
+            dist = dist_euclidean(y, y1)
+            if dist < eps:
+               break
+
+            y[:] = y1
+            iteration = iteration + 1
+            
+    return y
+
+cdef floating dist_euclidean(floating[:] x, floating[:] y) nogil:
+    cdef size_t n = x.shape[0]
+    cdef float64_t d = 0.
+    cdef float64_t tmp
+    for i in range(n):
+        tmp = x[i] - y[i]
+        d += tmp * tmp
+    return <floating>sqrt(d)
+
+cdef floating norm_euclidean(floating[:] x) nogil:
+    cdef size_t n = x.shape[0]
+    cdef float64_t d = 0.
+    for i in range(n):
+        d += x[i] * x[i]
+    return <floating>sqrt(d)
+
+cdef floating sum(floating[:] x) nogil:
```

Comment:
I don't know the purpose of this function, but I guess that it is for backward compatibility with float32. In our case we can simply do `x.sum()` (my guess; pending test).

Reply:
I'm not sure exactly what is going on as today is my first time working with Cython, but when I try using Python's built-in `sum()` function I get the following errors when I try to compile the Cython using `python setup.py build_ext --inplace`, all in reference to the `Dinvs = sum(Dinv)` line:
- `Converting to Python object not allowed without gil`
- `Constructing Python tuple not allowed without gil`
- `Calling gil-requiring function not allowed without gil`

[Here](https://cython.readthedocs.io/en/latest/src/userguide/nogil.html) is the documentation for the GIL. I don't think the Cython will compile correctly without this version of `sum()`.

Reply:
On a second thought, the originally design (together with the use of `nogil`) might be for parallelization.

Reply:
I believe that to be true, should we keep the original design then?